src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -94,90 +79,196 @@ public class DistributedTransactionContext : ITransactionContext, IEnlistmentNot
+					{
+						// In-doubt. A durable ressource has failed and may recover, but we won't wait to know.
+						RunAfterTransactionActions(false);
+					}
+
 					enlistment.Done();
-					IsInActiveTransaction = false;
 				}
 			}
 
-			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+			#endregion
+
+			public void TransactionCompleted(object sender, TransactionEventArgs e)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				e.Transaction.TransactionCompleted -= TransactionCompleted;
+				// This event may execute before second phase, so we cannot try to get the success from second phase.
+				// Using this event is required in case the prepare phase failed and called force rollback: no second
+				// phase would occur for this ressource.
+				var wasSuccessful = false;
+				try
 				{
-					logger.Debug("rolled back DTC transaction");
-					// Currently AfterTransactionCompletion is called by the handler for the TransactionCompleted event.
-					//sessionImplementor.AfterTransactionCompletion(false, null);
-					enlistment.Done();
-					IsInActiveTransaction = false;
+					wasSuccessful = e.Transaction.TransactionInformation.Status
+									== TransactionStatus.Committed;
 				}
+				catch (ObjectDisposedException ode)
+				{
+					_logger.Warn("Completed transaction was disposed, assuming transaction rollback", ode);
+				}
+				RunAfterTransactionActions(wasSuccessful);
 			}
 
-			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+			private volatile bool _afterTransactionActionDone;
+
+			private void RunAfterTransactionActions(bool wasSuccessful)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				if (_afterTransactionActionDone)
+					// Probably called from In-Doubt and TransactionCompleted.
+					return;
+				// Allow transaction completed actions to run while others stay blocked.
+				_bypassWait.Value = true;
+				try
 				{
-					sessionImplementor.AfterTransactionCompletion(false, null);
-					logger.Debug("DTC transaction is in doubt");
-					enlistment.Done();
-					IsInActiveTransaction = false;
+					using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+					{
+						// Flag active as false before running actions, otherwise the connection manager will refuse
+						// releasing the connection.
+						IsInActiveTransaction = false;
+						_sessionImplementor.AfterTransactionCompletion(wasSuccessful, null);
+						if (ShouldCloseSessionOnDistributedTransactionCompleted)
+						{
+							_sessionImplementor.CloseSessionFromDistributedTransaction();
+						}
+						_sessionImplementor.TransactionContext = null;
+					}
+				}
+				finally
+				{
+					_afterTransactionActionDone = true;
+					// Dispose releases blocked threads by the way.
+					// Must dispose in case !ShouldCloseSessionOnDistributedTransactionCompleted, since
+					// we nullify session TransactionContext, causing it to have nothing still holding it.
+					Dispose();
```

Comment:
This was missing and is a transaction leak in NHibernate 4.1. When scopes are disposed inside session life-span, the `TransactionContext` was nullified on session without being disposed, causing the `AmbientTransaction` clone to not be disposed of explicitly.